src/query-rule-context/__tests__/query-rule-context.spec.ts

```diff
@@ -0,0 +1,57 @@
+import { createRenderer } from '../../../helpers/test-renderer';
+import { NgAisQueryRuleContext } from '../query-rule-context';
+
+const defaultState = {
+  items: ['hello'],
+};
+
+describe('QueryRuleContext', () => {
+  let render;
+  let createWidget;
+  beforeEach(() => {
+    jest.clearAllMocks();
+    createWidget = jest.spyOn(NgAisQueryRuleContext.prototype, 'createWidget');
+    render = (template, state = defaultState) =>
+      createRenderer({
+        defaultState: state,
+        template,
+        TestedWidget: NgAisQueryRuleContext,
+      })({});
+  });
+
+  it('renders nothing without state', () => {
+    const fixture = render('<ais-query-rule-context></ais-query-rule-context>');
+
+    expect(fixture).toMatchSnapshot();
+  });
+
+  it('renders nothing with state', () => {
+    const fixture = render(
+      '<ais-query-rule-context></ais-query-rule-context>',
+      {
+        items: ['Luke', 'I am', 'your father'],
+      }
+    );
+
+    expect(fixture).toMatchSnapshot();
+  });
+
+  it('passes trackedFilters and transformRuleContexts', () => {
+    const trackedFilters = 'myTrackedFilters';
+    const transformRuleContexts = 'myTransformRuleContexts';
+    render(
+      `<ais-query-rule-context
+        [trackedFilters]="'${trackedFilters}'"
```

Comment:
This is the best workaround to test this to my knowledge. 
I wish there was an easier way to check functions are correctly passed.

Reply:
That's surprising that Angular doesn't provide the tools to test this. Besides, this doesn't help testing the type of the attributes.

Reply:
Angular doesn't respect the type on @Input anyway, because they come from HTML, it's really annoying actually